### ISSUE TITLE
HexTiles does behave strangely when used with hv.Dataset

### ISSUE THREAD

**Issue by anderl80** [holoviz/holoviews#3811]
I prepare a map using

```python
tile_opts = dict(height=plot_height, xaxis=None, yaxis=None, show_grid=False,
                 sizing_mode="scale_both", responsive=True)
map_tiles = gts.CartoLight.opts(style=dict(alpha=1), plot=tile_opts)
```
Then I do a plot with lon lat given in mercator coors

```python
points = (hv.HexTiles(data[["lon", "lat", "change"]], kdims=["lon", "lat"],
                     vdims=["value"])
          .options(active_tools=['wheel_zoom'],tools=[hover])
          .opts(colorbar=False, alpha=0.3, aggregator=np.mean, gridsize=250))

map_tiles * points
```

which works perfectly and plots the hexbins beautifully where they are located (Germany). However, as I am using a hv.Dataset (I want a dropdown with an additional key) like this

```python
points_ds = gv.Dataset(data[["lon", "lat", "value", "date"]])
points = (points_ds.to(hv.HexTiles, kdims=["lon", "lat"], vdims=["value"])
          .options(active_tools=['wheel_zoom'], tools=[hover])
          .opts(colorbar=False, alpha=0.3, gridsize=250))

map_tiles * points
```

the results are pretty strange. The plot does not appear most of times and when I click through the dropdown it's always on some other point in the map.

![image](https://user-images.githubusercontent.com/23196828/60896791-3c639080-a267-11e9-9be9-2fff4db45efa.png)

**Comment by philippjfr** [comment]
In the `.to` example are the coordinates in mercator coordinates or lat/lon? If the latter you should change it to:

```python
points = (points_ds.to(hv.HexTiles, kdims=["lon", "lat"], vdims=["value"])
          .options(active_tools=['wheel_zoom'], tools=[hover])
          .opts(colorbar=False, alpha=0.3, gridsize=250))
```

**Comment by anderl80** [comment]
all coords are projected to mercator. lat / lon is also mercator. Maybe I'm wrong, isn't your code example the same as above?

**Comment by philippjfr** [comment]
Yes, sorry, edited, I changed it to gv.HexTiles, but that won't help if it's already Mercator.

**Comment by poplarShift** [comment]
Just wanted to leave a pointer to https://github.com/pyviz/geoviews/issues/314, unsure if this issue is related to differences between hv.Dataset and gv.Dataset.

Also, it would be easier playing with this if you provided a Minimal Working Example.

**Comment by anderl80** [comment]
```python
from cartopy import crs as ccrs
import geoviews as gv
import holoviews as hv
import datashader as ds
from colorcet import bgy
import geoviews.tile_sources as gts
from holoviews.operation.datashader import datashade
gv.extension("bokeh")
import random
import numpy as np
import pandas as pd
import datetime

# map tiles
plot_width = 800
plot_height = 400
tile_opts = dict(height=plot_height, xaxis=None, yaxis=None, show_grid=False,
                 sizing_mode="scale_both", responsive=True)
map_tiles = gts.CartoLight.opts(style=dict(alpha=1), plot=tile_opts)

honolulu_coords = {"lat":21.455832, "lon":-157.974411}

def random_date(start, end):
    """Generate a random datetime between `start` and `end`"""
    return start + datetime.timedelta(
        # Get a random amount of seconds between `start` and `end`
        seconds=random.randint(0, int((end - start).total_seconds())),
    )
	
lon = [random.randint(0,100)/10000+honolulu_coords["lon"] for _ in range(0,1000)]
lat = [random.randint(0,100)/10000+honolulu_coords["lat"] for _ in range(0,1000)]
val = [random.randint(0,100) for _ in range(0,1000)]
date = [str(random_date(datetime.datetime(2019,1,1),datetime.datetime(2019,1,3)).date()) for _ in range(0,1000)]
data_gps = pd.DataFrame(list(zip(lon, lat, val, date)), columns=['lon', 'lat', 'val', 'date'])

# datashader and bokeh work with mercator proj
from datashader.geo import lnglat_to_meters
data_mercator = data_gps.copy(deep=True)
data_mercator.loc[:, 'lon'], data_mercator.loc[:, 'lat'] = lnglat_to_meters(data_mercator["lon"], data_mercator["lat"])

points = (hv.HexTiles(data_mercator[["lon", "lat", "val"]], kdims=["lon", "lat"],
                     vdims=["val"]).opts(aggregator=np.mean, gridsize=250))
					 
# plot 1
map_tiles * points

# plot 2 - same results as first plot
points_ds = gv.Dataset(data_gps[["lon", "lat", "val", "date"]])
map_tiles * points

# plot 3 - different result as the other plots
points_ds = gv.Dataset(data_mercator[["lon", "lat", "val", "date"]], crs=ccrs.Mercator())
map_tiles * points
```

The existance of lnglat_to_meters function suggests that is an appropriate function for this, but maybe it's the problem that it converts to web mercator and not mercator? But than it's strange that plot 1 and 2 are the same and plot 3 is different.

**Comment by anderl80** [comment]
Additionally, for my example (unfortunately I cannot share the raw data) I switched completely to gps (wgs84) coords in the dataframe.

```python
# beautifully shows the map
points = (gv.HexTiles(data_gps[["geohash_lon", "geohash_lat", "change"]], kdims=["geohash_lon", "geohash_lat"],
                     vdims=["change"])
          .options(active_tools=['wheel_zoom'],tools=[hover])
          .opts(colorbar=False, alpha=0.3, aggregator=np.mean, gridsize=250))
```
![image](https://user-images.githubusercontent.com/23196828/60956159-864d8480-a302-11e9-9b25-2c9480c7a4dc.png)

but

```python
points_ds = gv.Dataset(data_gps[["geohash_lon", "geohash_lat", "change", "date"]])
points = (points_ds.to(gv.HexTiles, kdims=["geohash_lon", "geohash_lat"], vdims=["change"])
          .options(active_tools=['wheel_zoom'], tools=[hover])
          .opts(colorbar=False, alpha=0.3, gridsize=250))
```

which is the same but I want a dropdown with the date, then it get's messed up

![image](https://user-images.githubusercontent.com/23196828/60956318-e17f7700-a302-11e9-8a0a-5f7cc8df294d.png)

**Comment by poplarShift** [comment]
Your example wasn't very minimal, but OK :) If I understand you correctly, your problem is that
1. `gv.Points(data_gps, ['lon', 'lat'])`, which is the same as `gv.Points(data_gps, ['lon', 'lat'], crs=ccrs.PlateCarree()) `, and 
2. `gv.Points(data_mercator, ['lon', 'lat'], crs=ccrs.Mercator())`

do not agree. That means that `lnglat_to_meters`  is different from the `ccrs.Mercator()` projection.

Is there any reason why you use the `lnglat_to_meters` function? Why not use the first option, where you are sure that it's lon/lat? Also, if you have to transform the lon/lat columns, is there a reason for not using `ccrs.Mercator().transform_points`?

**Comment by philippjfr** [comment]
``lnglat_to_meters`` is equivalent to ccrs.GOOGLE_MERCATOR I believe.

Separately though, there's no reason to manually project anything when you use GeoViews. So I'd recommend that you simply let GeoViews handle the projecting.

**Comment by anderl80** [comment]
Yes that's one thing that the crs do not seem to match, **but**, the problem exist with all crs, even when I use vanilla gps (wgs84) coordinates as suggestede with gv.
Everything does fine when using gv.HexTiles, gv.Points, ... **But**, when using a Dataset and the .to method, things go weird in combination with gv.HexTiles (with gv.Points, everything is fine).
To show this, I needed a bit more than a minimal example as seen above :-)

**Comment by anderl80** [comment]
> 
> 
> `lnglat_to_meters` is equivalent to ccrs.GOOGLE_MERCATOR I believe.
> 
> Separately though, there's no reason to manually project anything when you use GeoViews. So I'd recommend that you simply let GeoViews handle the projecting.

That's what I did in my second example https://github.com/pyviz/holoviews/issues/3811#issuecomment-509978746

data_gps are vanilla gps coordinates in lon/lat.

**Comment by poplarShift** [comment]
> Yes that's one thing that the crs do not seem to match, **but**, the problem exist with all crs, even when I use vanilla gps (wgs84) coordinates as suggestede with gv.
> Everything does fine when using gv.HexTiles, gv.Points, ... **But**, _when using a Dataset and the .to method, things go weird in combination with gv.HexTiles (with gv.Points, everything is fine)._
> To show this, I needed a bit more than a minimal example as seen above.

First, your map_tiles don't render for me with the options you give. So when I comment out the options, then your three plots have the same axis ranges for plot 1 and 2, but different for plot 3, and the axis ranges for each particular plot do not depend at all on whether I use Points or HexTiles.

Second, are the map_tiles really crucial for your minimal example? I cannot see a difference in axis ranges of the hextiles with or without.

Third, could you post a screenshot of what the below code gives for you?

![Screen Shot 2019-07-11 at 9 15 36 AM](https://user-images.githubusercontent.com/38992106/61053932-7e521b00-a3bc-11e9-9d51-f2aff2fa7d53.png)


```
from cartopy import crs as ccrs
import geoviews as gv
import holoviews as hv
import datashader as ds
from colorcet import bgy
import geoviews.tile_sources as gts
from holoviews.operation.datashader import datashade
gv.extension("bokeh")
import random
import numpy as np
import pandas as pd
import datetime

# map tiles
plot_width = 800
plot_height = 400
tile_opts = dict(height=plot_height, xaxis=None, yaxis=None, show_grid=False,
                 sizing_mode="scale_both", responsive=True)
map_tiles = gts.CartoLight#.opts(style=dict(alpha=1), plot=tile_opts)

honolulu_coords = {"lat":21.455832, "lon":-157.974411}

def random_date(start, end):
    """Generate a random datetime between `start` and `end`"""
    return start + datetime.timedelta(
        # Get a random amount of seconds between `start` and `end`
        seconds=random.randint(0, int((end - start).total_seconds())),
    )
	
lon = [random.randint(0,100)/10000+honolulu_coords["lon"] for _ in range(0,1000)]
lat = [random.randint(0,100)/10000+honolulu_coords["lat"] for _ in range(0,1000)]
val = [random.randint(0,100) for _ in range(0,1000)]
date = [str(random_date(datetime.datetime(2019,1,1),datetime.datetime(2019,1,3)).date()) for _ in range(0,1000)]
data_gps = pd.DataFrame(list(zip(lon, lat, val, date)), columns=['lon', 'lat', 'val', 'date'])

# datashader and bokeh work with mercator proj
from datashader.geo import lnglat_to_meters
data_mercator = data_gps.copy(deep=True)
data_mercator.loc[:, 'lon'], data_mercator.loc[:, 'lat'] = lnglat_to_meters(data_mercator["lon"], data_mercator["lat"])
				 
# plot 1
points_ds = gv.Dataset(data_mercator[["lon", "lat", "val", "date"]])
points = (points_ds.to(hv.HexTiles, kdims=["lon", "lat"], vdims=["val"])
          .opts(colorbar=False, alpha=0.3, gridsize=20))
p1 = map_tiles * points

# plot 2 - same results as first plot
points_ds = gv.Dataset(data_gps[["lon", "lat", "val", "date"]])
points = (points_ds.to(gv.HexTiles, kdims=["lon", "lat"], vdims=["val"])
          .opts(colorbar=False, alpha=0.3, gridsize=20))
p2 = map_tiles * points

# plot 3 - different result as the other plots
points_ds = gv.Dataset(data_mercator[["lon", "lat", "val", "date"]], crs=ccrs.Mercator())
points = (points_ds.to(gv.HexTiles, kdims=["lon", "lat"], vdims=["val"])
          .opts(colorbar=False, alpha=0.3, gridsize=20))
p3 = map_tiles * points

hv.output(p1)
hv.output(p2)
hv.output(p3)
```

**Comment by anderl80** [comment]
![image](https://user-images.githubusercontent.com/23196828/61122812-11fe1700-a4a3-11e9-9300-6292904a11c8.png)

**Comment by philippjfr** [comment]
So as long as I use ``ccrs.GOOGLE_MERCATOR`` instead of ``ccrs.Mercator()`` I can't reproduce any issues with your example.

**Comment by anderl80** [comment]
So maybe we should now try to add a map, maybe this is related with the map.

Still my example with using raw gps coordinates on combination with geoviews and no declaration of a ref crs does not work. I will redo the example later.

I have problems with raw gps coordinates (wgs84) and giving no crs in a few other examples and gv. Got an error related to PlateCaree...

**Comment by poplarShift** [comment]
So it looks like everything is fine then, right? 

(Of course not giving the `crs` argument may lead to wrong results.)

**Comment by anderl80** [comment]
No, see the message above, the original problem still persists.

Now we got the (a) crs thing solved (what is the crs for wgs84?), although I thought that plain gps coordinates do not need crs explicitly referenced. (b) we got the thing solved without using a background map. However if we use the map then the original problem persists.

I will give an updated code example tomorrow.

**Comment by poplarShift** [comment]
I see, sorry, I got sidetracked by the CRS issue. (CRS for plain lat/lon is indeed PlateCarree.)

**Comment by anderl80** [comment]
No worries, thanks for the support and sorry for distracting from the main problem.

I was able do "extract"  (still sorry for the over-engineered example :-)) the problem more. Please try to execute the following.

```python
# generate random data points for GitHub issue
import random
import datetime
import numpy as np
import pandas as pd

muc_coords = {"lat": 48.137154, "lon": 11.576124}

# generate data
def random_date(start, end):
    """Generate a random datetime between `start` and `end`"""
    return start + datetime.timedelta(
        # Get a random amount of seconds between `start` and `end`
        seconds=random.randint(0, int((end - start).total_seconds())),
    )
lon = [random.randint(0,100)/10000+muc_coords["lon"] for _ in range(0,1000)]
lat = [random.randint(0,100)/10000+muc_coords["lat"] for _ in range(0,1000)]
val = [random.randint(0,100) for _ in range(0,1000)]
date = [str(random_date(datetime.datetime(2019,1,1),datetime.datetime(2019,1,10)).date()) for _ in range(0,1000)]
data = pd.DataFrame(list(zip(lon, lat, val, date)), columns=['lon', 'lat', 'change', 'date'])

# [x] check that there's no day without data
data.date.value_counts().sort_index()

import geoviews as gv
gv.extension("bokeh")

hexs = (gv.HexTiles(data, kdims=["lon", "lat"],
                     vdims=["change", "date"])
          .options(active_tools=['wheel_zoom'],tools=["hover"])
          .opts(colorbar=False, alpha=0.3, aggregator=np.mean))
		  
# everything works fine
hexs

# now we try to add a dropdown menu
hexs_ds = gv.Dataset(data)

hexs = (hexs_ds.to(gv.HexTiles, kdims=["lon", "lat"], vdims=["change"])
          .options(active_tools=['wheel_zoom'], tools=["hover"])
          .opts(colorbar=False, alpha=0.3))
		  
# 1. click through dropdown menu and find shifted data (zoom out)
# 2. also find hex are not aggregated over "change" (vdim) but "count"
hexs
```

#1 gives me
![image](https://user-images.githubusercontent.com/23196828/61452851-086e2680-a95d-11e9-9fc8-4867929bbdec.png)

#2 gives me (already zoomed out because the starting graph was empty)
![image](https://user-images.githubusercontent.com/23196828/61452902-25a2f500-a95d-11e9-8f88-ffce60e75c06.png)

**Comment by philippjfr** [comment]
Appears to have been fixed in #4182
